### PR TITLE
feat(ssr): vue-ssr-webpack-plugin compatible with webpack 5

### LINKED ISSUE #11718: vuejs/vue-ssr-webpack-plugin and webpack 5
### What problem does this feature solve?
github missing repository vuejs/vue-ssr-webpack-plugin, i update webpack to version 5 and have error
[vue-server-renderer-webpack-plugin] webpack config `output.libraryTarget` should be "commonjs2".

### What does the proposed API look like?
in my output.libraryTarget is libraryTarget: 'commonjs2'

### PR DESCRIPTION
**Pull request by gzzhanghao**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

Similar to #11863, but:

- Fix modules mapping in client plugin
- Ignore hmr assets
- Fix the deprecation warning

Fix #11718
Close #11863

### CODE REVIEW

**Comment by redneckz** [comment]
Hi colleagues.
Have a nice day.

The last activity was several weeks ago...
Any progress on this issue?

Currently, we try to move to Webpack 5 and have troubles with Vue SSR plugin.
We stuck trying to migrate to micro frontend architecture (by means of Module Federation). In our case, it is a blocker :(

Please give some roadmap to adjust our deadlines and make appropriate decisions.

Thanks in advance.

**Comment by gzzhanghao** [comment]
@redneckz It's not a good practice, but you can copy `server-plugin.js` and `client-plugin.js` to your project and use it locally.

https://github.com/gzzhanghao/vue/tree/ssr-webpack-5-build/packages/vue-server-renderer

File: src/server/webpack-plugin/server.js [hunk 1]
```diff
@@ -10,7 +10,8 @@ export default class VueSSRServerPlugin {
   apply (compiler) {
     validate(compiler)
 
-    onEmit(compiler, 'vue-server-plugin', (compilation, cb) => {
+    const stage = 'PROCESS_ASSETS_STAGE_OPTIMIZE_TRANSFER'
```
Comment (sodatea) [inline src/server/webpack-plugin/server.js:13]: Why not just put the stage name in the `onEmit` function? It's not used anywhere else

**Review by sodatea — CHANGES_REQUESTED** [review]

File: src/server/webpack-plugin/server.js (hunk 1 above)
Comment (gzzhanghao) [reply src/server/webpack-plugin/server.js:13]: @sodatea client plugin and server plugin are applied in different stage, so I think it's better to specify the stage name here.

**Review by sodatea — APPROVED** [review]

**Pull request merged**